File: services/order-service/src/main/java/com/ticketing/orders/outbox/OutboxRelay.java
package com.ticketing.orders.outbox;

import com.ticketing.orders.entity.OutboxMessage;
import com.ticketing.orders.repository.OutboxRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.domain.Pageable;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Polls the outbox table at a configurable interval and publishes unpublished messages to Kafka.
 * Default poll interval is 5 000 ms — override via OUTBOX_RELAY_POLL_INTERVAL_MS env var.
 * Reducing from 500 ms to 5 000 ms cuts DB queries and OTel spans ~10× while keeping
 * end-to-end event latency well within the 15-minute order expiry window.
 *
 * Each poll claims a bounded page (default 50, OUTBOX_RELAY_BATCH_SIZE) rather than the whole
 * unpublished backlog, so a Kafka outage cannot turn the backlog into an OOM.
 *
 * The claim uses FOR UPDATE SKIP LOCKED, which only isolates replicas for as long as the
 * claiming transaction lives — hence {@code @Transactional} on this method. That supersedes
 * the earlier per-message-transaction arrangement (C-02): with 2–8 replicas, per-message
 * commits meant every replica published every row on every poll, which is a far larger
 * correctness problem than the batch commit this reintroduces. Batch scope is bounded by the
 * page size.
 *
 * On a failed send the batch stops at the failing row and commits the rows already published;
 * the failing row and everything after it stay unpublished and are retried next poll. Stopping
 * rather than skipping is what preserves per-entity ordering: skipping would let a later event
 * for the same partition key reach Kafka ahead of an earlier one that is still failing, which
 * defeats the point of keying by orderId at all (AGENTS.md §3.4). The cost is head-of-line
 * blocking if a row can never be published (e.g. a payload over max.message.bytes); that shows
 * up as a repeating publish-failure log rather than as silent reordering. This matches
 * payment-service and attendance-service, which make the same trade-off.
 *
 * A DB failure while marking, by contrast, marks the batch rollback-only, so the whole page is
 * re-published — bounded by the page size, and consumers must be idempotent regardless
 * (AGENTS.md §3.5).
 *
 * The partition key stored in the outbox is used as the Kafka message key so that
 * messages for the same entity (e.g. same orderId) land on the same partition,
 * preserving per-entity ordering (AGENTS.md §3.4).
 */
@Component
public class OutboxRelay {

    private static final Logger log = LoggerFactory.getLogger(OutboxRelay.class);

    private final OutboxRepository outboxRepository;
    private final OutboxMessagePublisher publisher;
    private final int batchSize;

    public OutboxRelay(OutboxRepository outboxRepository,
                       OutboxMessagePublisher publisher,
                       @Value("${outbox.relay.batch-size:50}") int batchSize) {
        this.outboxRepository = outboxRepository;
        this.publisher = publisher;
        this.batchSize = batchSize;
    }

    @Scheduled(fixedDelayString = "${outbox.relay.poll-interval-ms:5000}")
    @Transactional
    public void relay() {
        List<OutboxMessage> pending = outboxRepository.findUnpublished(Pageable.ofSize(batchSize));
        if (pending.isEmpty()) {
            return;
        }

        int published = 0;
        for (OutboxMessage msg : pending) {
            if (!publisher.publishOne(msg)) {
                log.warn("Outbox relay stopped at message id={} topic={}; committing {} row(s) already "
                                + "published and retrying from this row next poll",
                        msg.getId(), msg.getTopic(), published);
                break;
            }
            published++;
        }
    }
}

File: services/attendance-service/internal/service/outbox_relay.go
package service

import (
	"context"
	"fmt"
	"time"

	"github.com/acme/attendance-service/internal/repository"
	"github.com/jackc/pgx/v5"
	"go.uber.org/zap"
)

// TxBeginner is satisfied by *pgxpool.Pool and by test doubles.
type TxBeginner interface {
	Begin(ctx context.Context) (pgx.Tx, error)
}

// OutboxRelay publishes queued outbox rows and marks them as published.
// Each RunOnce call opens a single transaction and claims rows with
// FOR UPDATE SKIP LOCKED so that concurrent relay replicas (HPA scale-out)
// each receive disjoint sets of rows and never double-publish.
type OutboxRelay struct {
	db   TxBeginner
	repo repository.OutboxRepository
	pub  EventPublisher
	log  *zap.Logger
}

func NewOutboxRelay(db TxBeginner, repo repository.OutboxRepository, pub EventPublisher, log *zap.Logger) *OutboxRelay {
	return &OutboxRelay{db: db, repo: repo, pub: pub, log: log}
}

func (r *OutboxRelay) Run(ctx context.Context, interval time.Duration, batchSize int) {
	ticker := time.NewTicker(interval)
	defer ticker.Stop()

	for {
		if err := r.RunOnce(ctx, batchSize); err != nil {
			r.log.Error("attendance outbox relay iteration failed", zap.Error(err))
		}

		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
		}
	}
}

func (r *OutboxRelay) RunOnce(ctx context.Context, batchSize int) error {
	// At-least-once semantics. Rows are claimed inside a transaction with
	// FOR UPDATE SKIP LOCKED (via ListUnpublishedTx); the claim is held for the
	// whole batch, which is what keeps concurrent replicas disjoint.
	//
	// A publish failure at row N stops the batch but does NOT discard the batch:
	// rows 1..N-1 are already on the topic, so their marks are committed. Rolling
	// them back would re-send every one of them next tick — a guaranteed burst of
	// duplicates on every partial failure, growing with batch position. Row N is
	// left unmarked because we cannot tell whether it reached the broker, so it
	// is retried. Downstream consumers MUST still be idempotent on credential_id.
	//
	// Stopping rather than skipping row N preserves per-entity ordering: skipping
	// would let a later event for the same partition key reach Kafka ahead of an
	// earlier one still being retried.
	tx, err := r.db.Begin(ctx)
	if err != nil {
		return fmt.Errorf("relay: begin tx: %w", err)
	}
	defer func() { _ = tx.Rollback(ctx) }()

	rows, err := r.repo.ListUnpublishedTx(ctx, tx, batchSize)
	if err != nil {
		return fmt.Errorf("relay: list unpublished outbox rows: %w", err)
	}

	var publishErr error
	for _, row := range rows {
		if err := r.pub.Publish(row.Topic, []byte(row.PartitionKey), row.Payload); err != nil {
			r.log.Warn("relay: publish failed; committing progress and retrying this row next tick",
				zap.String("id", row.ID), zap.Error(err))
			publishErr = fmt.Errorf("relay: publish outbox row %s: %w", row.ID, err)
			break
		}
		if err := r.repo.MarkPublishedTx(ctx, tx, row.ID, time.Now().UTC()); err != nil {
			// Unlike a publish failure this aborts the transaction, so there is
			// no progress left to commit — every row in the batch is re-published.
			return fmt.Errorf("relay: mark outbox row %s published: %w", row.ID, err)
		}
	}

	if err := tx.Commit(ctx); err != nil {
		return fmt.Errorf("relay: commit outbox batch: %w", err)
	}
	return publishErr
}
